keys/signer_ed25519.go

```diff
@@ -0,0 +1,82 @@
+// Copyright 2020 Coinbase, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package keys
+
+import (
+	"crypto/ed25519"
+	"encoding/hex"
+	"fmt"
+
+	"github.com/coinbase/rosetta-sdk-go/asserter"
+	"github.com/coinbase/rosetta-sdk-go/types"
+)
+
+// SignerEd25519 is initialized from a keypair
+type SignerEd25519 struct {
+	KeyPair KeyPair
+}
+
+// PublicKey returns the PublicKey of the signer
+func (s SignerEd25519) PublicKey() types.PublicKey {
+	return *s.KeyPair.PublicKey
+}
+
+// Signs arbitrary payloads using a KeyPair
+func (s SignerEd25519) Sign(payload types.SigningPayload) (*types.Signature, error) {
+	err := s.KeyPair.IsValid()
+	if err != nil {
+		return nil, err
+	}
+	privKeyBytes := s.KeyPair.PrivateKey.Bytes
+	decodedMessage, err := hex.DecodeString(payload.HexBytes)
+	if err != nil {
+		return nil, fmt.Errorf("sign: unable to decode message. %w", err)
+	}
+
+	privKey := ed25519.NewKeyFromSeed(privKeyBytes)
+	sig := ed25519.Sign(privKey, decodedMessage)
+
+	return &types.Signature{
+		SigningPayload: &payload,
+		PublicKey:      s.KeyPair.PublicKey,
+		SignatureType:  payload.SignatureType,
+		HexBytes:       hex.EncodeToString(sig),
+	}, nil
+}
+
+// Verify verifies a Signature, by checking the validity of a Signature,
+// the SigningPayload, and the PublicKey of the Signature.
+func (s SignerEd25519) Verify(signature *types.Signature) (bool, error) {
+	pubKey, err := hex.DecodeString(signature.PublicKey.HexBytes)
+	if err != nil {
+		return false, fmt.Errorf("verify: unable to decode pubkey. %w", err)
+	}
+	decodedMessage, err := hex.DecodeString(signature.SigningPayload.HexBytes)
+	if err != nil {
+		return false, fmt.Errorf("verify: unable to decode message. %w", err)
+	}
+	decodedSignature, err := hex.DecodeString(signature.HexBytes)
```

Comment:
Is it possible to also update `types.Signature`, `types.SigningPayload` and `types.PublicKey` type to use `[]byte` instead so you can avoid all this decoding?

Since these already exist I'd expect this to potentially have larger implications and we could follow up with it, but I think the amount of extra decoding necessary shows that the way we're storing this data doesn't match how it's used, so making that closer could be beneficial.

Reply:
@patrick-ogrady thoughts here?

Reply:
If each signer implementation takes `[]byte` instead of these Rosetta types, you would only need to implement the decoding logic once in some wrapper (which has the Rosetta types as args) around any signer implementation. This approach could lead to less redundant encoding/decoding code and make the signer implementations more minimal (probably preferred).

Still thinking about the best format for the wrapper but a naive approach could be something as simple as taking a signer as the first argument in a collection of functions with Rosetta types (the same functions on the Signer interface).

How do you feel about this @juliankoh?

Reply:
> This approach could lead to less redundant encoding/decoding code and make the signer implementations more minimal (probably preferred).

Reply:
I don't understand enough about the SDK structure to comment much on using a wrapper type or not, but I'd strongly recommend against passing around hex strings as much as possible. 

Each one is an opportunity to have a bug in decoding logic and makes getting a correct implementation more difficult. When we're decoding things like private keys, it's even more important to make sure we don't make mistakes, so requiring extra effort to decode is a risk. If we use the strong types to our advantage and use `[]byte` instead in every place _except the API layer_ where we can't because of JSON, we end up with a safer system that's easier to integrate with and less likely to have bugs.

Reply:
I resolved this issue in #60. `types.Signature`, `types.SigningPayload`, and `types.PublicKey` all now use `[]byte`! Thanks for the feedback @jakecraige.